test/comparison-tests/projectReferencesWatchRefWithTwoFilesAlreadyBuilt_Composite_WatchApi/expectedOutput-3.9/patch2/output.txt

```diff
@@ -1,7 +1,8 @@
-    Asset      Size  Chunks             Chunk Names
- app.d.ts  11 bytes          [emitted]  
-bundle.js  4.85 KiB    main  [emitted]  main
+               Asset      Size  Chunks             Chunk Names
```

Comment:
This change seems incorrect from project references point of view. It should not write lib\tsconfig.tsbuildinfo file since there are errors in the build of project `lib` this will cause unknown issues as project cannot have tsbuild info file if there are errors. (not in 3.9 for sure)

Reply:
Thanks for commenting @sheetalkamat!  Would you advise reverting this?

In a side note, @appzuka has been observing that the watch API support (which I believe you originally added 🤗) does not improve performance of `ts-loader`. This is surprising given that it provides a massive performance boost to `fork-ts-checker-webpack-plugin` - however that does seem to be the case.

I'd always hoped that the watch API usage would become the default way to run ts-loader. The idea being that it stayed behind the `experimentalWatchApi` flag until it was fully functional and out the servicehost approach. However, that day has not come sadly! Due to various bugs and the performance issue we're now thinking about potentially removing it.

What are your thoughts? I'm mindful you put a good amount of effort into implementation in the first place (which we're super grateful for BTW!) And also, if there's a simple thing `ts-loader` is doing slightly wrong here that would make the watch API perform I'd still love to make that happen. Again I can't forget that `fork-ts-checker-webpack-plugin` got turbo charged from using the watch API so there's a good chance we're doing something sub optimal.

What do you think?

cc @andrewbranch in case there's anything he'd like to add to this discussion too

Reply:
Looking at the commit prior to this PR, in projectReferencesWatchRefWithTwoFilesAlreadyBuilt, patch2 is supposed to use the new number added to the helper in patch 1:

https://github.com/TypeStrong/ts-loader/blob/a77470f9072bf24dc7165653c2ee86b243b65040/test/comparison-tests/projectReferencesWatchRefWithTwoFilesAlreadyBuilt/patch2/app.ts#L4

In the output bundle you can see the new number has been added:

https://github.com/TypeStrong/ts-loader/blob/a77470f9072bf24dc7165653c2ee86b243b65040/test/comparison-tests/projectReferencesWatchRefWithTwoFilesAlreadyBuilt/expectedOutput-3.9/patch2/bundle.js#L97

But in the watchAPI version it does not:

https://github.com/TypeStrong/ts-loader/blob/a77470f9072bf24dc7165653c2ee86b243b65040/test/comparison-tests/projectReferencesWatchRefWithTwoFilesAlreadyBuilt_WatchApi/expectedOutput-3.9/patch2/bundle.js#L97

And the same with the Composite_WatchApi

https://github.com/TypeStrong/ts-loader/blob/a77470f9072bf24dc7165653c2ee86b243b65040/test/comparison-tests/projectReferencesWatchRefWithTwoFilesAlreadyBuilt_Composite_WatchApi/expectedOutput-3.9/patch2/bundle.js#L97

Also note that tsconfig.tsbuildinfo is not written to the patch2 output:

https://github.com/TypeStrong/ts-loader/tree/a77470f9072bf24dc7165653c2ee86b243b65040/test/comparison-tests/projectReferencesWatchRefWithTwoFilesAlreadyBuilt_Composite_WatchApi/expectedOutput-3.9/patch2

After the [PR#1159](https://github.com/TypeStrong/ts-loader/pull/1159) the output of patch2 correctly includes the new content and the tsconfig.tsbuildinfo:

https://github.com/TypeStrong/ts-loader/blob/2f6cbe84777e07364f89be5db97d1a02358554c5/test/comparison-tests/projectReferencesWatchRefWithTwoFilesAlreadyBuilt_Composite_WatchApi/expectedOutput-3.9/patch2/bundle.js#L97

Note that the new tsconfig.tsbuildinfo is in the top level directory not in the lib folder.

I was concerned that the PR changed the comparison test results but I believe the previous output was incorrect so the changes reflect the fixing of behaviour by the PR.  The watchapi code is quite complex so there could be further issues which need to be fixed.

Reply:
@appzuka my bad.. i saw that tsbuildinfo as from lib project instead of root project. Change looks good.

Reply:
@johnnyreilly Yes watchAPI should be faster and should see benefit.. I am investigating performance for internal team that uses ts-loader with solution builder work.. i can take a issue with watch API but i will need concrete repro steps so i can investigate. Watch API is kind of similar to solution builder API (how it handles file changes etc) so current perf work i am doing might help not sure. In the mean time if you can get repro where not using watch api is faster than watch api, i will be better equipped to investigate it. without that its just analysing code which i might not have enough time to do.

Reply:
@sheetalkamat - thanks so much!

@appzuka you've bumped on the problems with the performance of the watch API the most as you've been looking into this. I'm bearing in mind your comment here:

> This PR fixes these issues. When I use this on a real project however `experimentalWatchApi` is very slow. Without `experimentalWatchApi` the build time for a larger project is around 37 seconds but with the watchApi it jumps to 168 seconds. This PR increases the build time to 199 seconds. Using transpileOnly and for-ts-checker the build time is reduced to 23 seconds.

Would you be up for providing some kind of demo repo which demonstrates performance being better with the non watch API vs the watch API?

No worries if you can't - I can try and get to it later (after my holidays)